### PR TITLE
First cut at using generics for base flag types

### LINKED ISSUE #833: Release 3.x meta/discussion
This is a meta-issue meant to track the work in _other issues_ targeting the [Release 3.x milestone](https://github.com/urfave/cli/milestone/5).

<details>
  <summary>benevolent takeover notes</summary>

Hiii I'm working on getting many maintenance tasks addressed including
taking over this issue from @coilysiren 🙇🏼 who is busy with other life adventures ❤️ ~ @meatballhat

</details>

~I know 2.0 isn't out yet 🙂 (see https://github.com/urfave/cli/issues/826) but~ I wanted to create a tracking issue for the 3.x series.

Here's some breaking/difficult changes we want to consider for V3:

- [x] #1498 
- [x] #1489 
- [x] Replace code generation with generics where possible/maintainable (#1512 )
- [x] ~#1058~ 
- [x] #1055 
- [x] #1695 
- [x] ~Improve testability of `*cli.Context` (via https://github.com/urfave/cli/issues/833#issuecomment-574171962)~ (conflicts with #1587)
- [x] remove `EnableBashCompletion` in favor of `CompletionShells`
- [x] ~consider a different release method (eg. not just merging to `master`) via https://github.com/urfave/cli/issues/952#issuecomment-561062742~ (this was changed in v2)
- [x] ~make flags "global" by default, to some degree~ (this was added into v2)
- [x] #753 
- [x] #1071
- [x] ~Cleanup error handling. Make multiError implement ExitCoder~
- [x] #1663
- [x] #1890 [^1]
- [x] #1775
- [x] #1531
- [x] #1891

Please suggest more! I'll add them to this OP

**Update ~January 2023**: Progress! And yet, yeah, wish lists balanced with releasing usable software is hard to do. If you're the kind of human who finds themselves reading this, there's a `v3.0.0-alpha2` tag now, which should play nicely with `go get github.com/urfave/cli/v3@latest`. Feedback wanted! ❤️ ~ @meatballhat 

[^1]: The documentation-related code that depends on `github.com/russross/blackfriday/v2` and `github.com/cpuguy83/go-md2man/v2` has been [shown to add significant bulk to compiled outputs](https://github.com/urfave/cli/pull/1375#issuecomment-1109117786)

### PR DESCRIPTION
**Pull request by dearchap**
## What type of PR is this?

_(REQUIRED)_

- cleanup

## What this PR does / why we need it:

_(REQUIRED)_

Use generics all the way

## Which issue(s) this PR fixes:

_(REQUIRED)_

Fixes #833 
## Special notes for your reviewer:

_(fill-in or delete this section)_

## Testing

_(fill-in or delete this section)_

## Release Notes

_(REQUIRED)_

```release-note

```

### CODE REVIEW

**Comment by jtagcat** [comment]
iirc using generics lessens type checking, and moves errors to runtime, but it might work as a wrapper to a flag lib itslef

**Comment by jtagcat** [comment]
What were you experimenting with `golang.org/x/exp`?

**Comment by dearchap** [comment]
> 

Not sure, it wasnt a direct import

**Comment by jtagcat** [comment]
I can't yet put my finger on it, but be aware, that multiple interface typing (`interface{}.(struct).(int)`) will panic if the larger struct is not found, under a `if v, ok := `.

`if v, ok := ` resolves panic / gives ok value only on the last `.(int)`. If anything before it is not found, it will panic.

**Comment by jtagcat** [comment]
jfyi generics are not the keyword here, what you/we are doing is implementing/using (generic) interfaces

**Comment by jtagcat** [comment]
I'd move flags to a separate subpackage (a subdirectory), this would most importantly isolate it, but also enable possible external uses.

File: app_test.go [hunk 1]
```diff
@@ -2654,6 +2654,7 @@ func TestFlagAction(t *testing.T) {
 	stringFlag := &StringFlag{
 		Name: "f_string",
 		Action: func(c *Context, v string) error {
+			t.Log("in adction")
```
[the remark this answers is not in the document]
Comment (dearchap) [reply app_test.go:2729]: Yeah that was a debug log. Removed now.

File: docs_test.go [hunk 2]
```diff
@@ -1,5 +1,5 @@
-//go:build !urfave_cli_no_docs
-// +build !urfave_cli_no_docs
+//go:build to_fix
+// +build to_fix
```
Comment (dearchap) [inline docs_test.go:2]: @meatballhat i need your help to fix this

Comment (meatballhat) [reply docs_test.go:2]: It looks to me like all of the test diffs are around bool flags having usage text with `(default: false)`. If that's an intentional change, then I think what's needed is for the `testdata/expected-*` files to be updated accordingly.

Comment (dearchap) [reply docs_test.go:2]: Ok thanks let me take a look.

File: flag_string.go [hunk 3]
```diff
@@ -2,75 +2,30 @@ package cli
...
 	"flag"
-	"fmt"
 )
 
-// GetValue returns the flags value as string representation and an empty
-// string if the flag takes no value at all.
-func (f *StringFlag) GetValue() string {
-	return f.Value
-}
+// -- string Value
```
Comment (meatballhat) [inline flag_string.go:7]: Is there special significance to the `// --`?

File: sliceflag.go [hunk 4]
```diff
@@ -229,7 +229,7 @@ func unwrapFlagValue(v flag.Value) flag.Value {
 
 // NOTE: the methods below are in this file to make use of the build constraint
 
-func (f *Float64SliceFlag) SetValue(slice []float64) {
+/*func (f *Float64SliceFlag) SetValue(slice []float64) {
```
Comment (meatballhat) [inline sliceflag.go:232]: Did you want to address these commented-out slice type bits within this PR or leave it as later work?

**Review by meatballhat — APPROVED** [review]

File: flag_impl.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+type FlagConfig interface {
+	IntBase() int
+	GetCount() *int
+}
+
+type ValueCreator[T any] interface {
+	Create(T, *T, FlagConfig) flag.Value
+}
+
+// Float64Flag is a flag with type float64
```
[the remark this answers is not in the document]
Comment (dearchap) [reply flag_impl.go:18]: Updated

File: flag_impl.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+	hasBeenSet bool
+
+	Value       T
+	Destination *T
+
+	Aliases []string
+	EnvVars []string
+
+	TakesFile bool
+	Base      int
```
[the remark this answers is not in the document]
Comment (dearchap) [reply flag_impl.go:38]: yes makes sense

File: flag_string.go (hunk 3 above)
Comment (dearchap) [reply flag_string.go:7]: ha this is cut and paste from the golang flag code

File: sliceflag.go (hunk 4 above)
Comment (dearchap) [reply sliceflag.go:232]: I will remove these commented out slice types

File: flag_slice_impl.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+// NewIntSlice makes an *IntSlice with default values
+func NewSliceBase[T any, C any, VC ValueCreator[T, C]](defaults ...T) *SliceBase[T, C, VC] {
+	return &SliceBase[T, C, VC]{
+		slice: &defaults,
+	}
+}
+
+// TODO: Consistently have specific Set function for Int64 and Float64 ?
+// SetInt directly adds an integer to the list of values
+func (i *SliceBase[T, C, VC]) SetOne(value T) {
```
Comment (meatballhat) [inline flag_slice_impl.go:38]: This looks resolved now (?)

File: flag_slice_impl.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+	if !i.hasBeenSet {
+		*i.slice = []T{}
+		i.hasBeenSet = true
+	}
+
+	*i.slice = append(*i.slice, value)
+}
+
+// Set parses the value into an integer and appends it to the list of values
+func (i *SliceBase[T, C, VC]) Set(value string) error {
```
Comment (meatballhat) [inline flag_slice_impl.go:48]:
```suggestion
// Set parses the value and appends it to the list of values
func (i *SliceBase[T, C, VC]) Set(value string) error {
```

File: flag_slice_impl.go [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+	v := i.Value()
+	var t T
+	if reflect.TypeOf(t).Kind() == reflect.String {
+		return fmt.Sprintf("%v", v)
+	}
+	return fmt.Sprintf("%T{%s}", v, i.ToString(v))
+}
+
+// Serialize allows IntSlice to fulfill Serializer
+func (i *SliceBase[T, C, VC]) Serialize() string {
```
Comment (meatballhat) [inline flag_slice_impl.go:86]:
```suggestion
// Serialize allows SliceBase to fulfill Serializer
func (i *SliceBase[T, C, VC]) Serialize() string {
```

File: flag_slice_impl.go [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+	jsonBytes, _ := json.Marshal(i.slice)
+	return fmt.Sprintf("%s%s", slPfx, string(jsonBytes))
+}
+
+// Value returns the slice of ints set by this flag
+func (i *SliceBase[T, C, VC]) Value() []T {
```
Comment (meatballhat) [inline flag_slice_impl.go:92]:
```suggestion
// Value returns the slice of values set by this flag
func (i *SliceBase[T, C, VC]) Value() []T {
```

File: flag_timestamp.go [hunk 11]
```diff
@@ -52,114 +62,26 @@ func (t *Timestamp) Set(value string) error {
...
-
-// RunAction executes flag action if set
-func (f *TimestampFlag) RunAction(c *Context) error {
-	if f.Action != nil {
-		return f.Action(c, c.Timestamp(f.Name))
-	}
-
-	return nil
+func (t *timestampValue) Get() interface{} {
+	return *t.timestamp
```
Comment (meatballhat) [inline flag_timestamp.go:84]: Does this method still have to return `interface{}`?

File: flag_uint.go [hunk 12]
```diff
@@ -1,84 +1,52 @@
...
-		return f.Action(c, c.Uint(f.Name))
-	}
+func (i *uintValue) Get() any { return uint(*i.val) }
 
-	return nil
-}
+func (i *uintValue) String() string { return strconv.FormatUint(uint64(*i.val), 10) }
 
-// Uint looks up the value of a local UintFlag, returns
+// Int looks up the value of a local IntFlag, returns
```
Comment (meatballhat) [inline flag_uint.go:45]:
```suggestion
// Uint looks up the value of a local UintFlag, returns
```

File: flag_uint64.go [hunk 13]
```diff
@@ -1,84 +1,52 @@
...
-		return f.Action(c, c.Uint64(f.Name))
-	}
+func (i *uint64Value) Get() any { return uint64(*i.val) }
 
-	return nil
-}
+func (i *uint64Value) String() string { return strconv.FormatUint(uint64(*i.val), 10) }
 
-// Uint64 looks up the value of a local Uint64Flag, returns
+// Int64 looks up the value of a local Int64Flag, returns
```
Comment (meatballhat) [inline flag_uint64.go:45]:
```suggestion
// Uint64 looks up the value of a local Uint64Flag, returns
```

**Review by meatballhat — APPROVED** [review]
nits only! This is very exciting! 🎉

File: flag_slice_impl.go (hunk 7 above)
Comment (dearchap) [reply flag_slice_impl.go:38]: yes.

File: flag_timestamp.go (hunk 11 above)
Comment (dearchap) [reply flag_timestamp.go:84]: Yeah that is to satisfy the flag.Getter.Get() any function.

**Pull request merged**